### ISSUE TITLE
High disk I/O prevents successful DNS resolution

### ISSUE THREAD

**Issue by nazar-pc** [pi-hole/pi-hole#5635]
### Versions

- Pi-hole:  v5.18.2
- AdminLTE:  v5.21
- FTL:  v5.25.1

### Platform

- OS and version:  Ubuntu 22.04

- Platform:  Docker

### Expected behavior

DSN resolution should work all the time

### Actual behavior / bug

On very high disk I/O DNS resolution simply stops working. For example this happens when running `sync` on NVMe SSD after large disk operation that on modern Linux with `noop` I/O scheduler basically blocks a lot of things for a significant amount of time.

When Pi-hope recovers I see this in logs:
```
Long-term load (15min avg) larger than number of processors: 80.3 > 32
This may slow down DNS resolution and can cause bottlenecks.
```

### Steps to reproduce

Have Pi-hope runing on disk with very high I/O load and observe DNS resolution failing.

## Debug Token

https://tricorder.pi-hole.net/sgxHNmzd/

### Additional context

My educated guess is that on each DNS request Pi-hole tries to read something from disk and under high I/O read times out.

Considering how small gravity database is (8.7M in my case), it would make a lot of sense to keep its copy in memory, something under 50M will not cause issues for most users, while performance will improve massively and will avoid issue described here completely.

**Comment by rdwebdesign** [comment]
Pi-hole database is already kept in memory and new queries are written to disk from time to time. 
If not possible to write immediately, Pi-hole will try to write the data later.

**Comment by nazar-pc** [comment]
Interesting, then my guess is incorrect, but the issue is still present. How can I debug it to provide more information?

**Comment by rdwebdesign** [comment]
Please, generate and upload a Debug Log:

**Using the command line:**
- use `pihole -d` command.
- when asked to upload the log, answer <kbd>Y</kbd>.
- copy only the token and paste here.

**Using the web interface:**
- click on _**Tools > Generate Debug Log**_ menu.
- then mark the second checkbox to upload the log and click on the big button.
- when finished, copy only the token and paste here.

**Comment by nazar-pc** [comment]
https://tricorder.pi-hole.net/sgxHNmzd/

**Comment by rdwebdesign** [comment]
I'm not 100% sure the issue is related to high I/O load causing a Pi-hole failure.

Usually, when it fails to write to the database you see messages like these:

```
ERROR: SQL query "END TRANSACTION" failed: database is locked (SQLITE_BUSY)
END TRANSACTION failed when trying to store queries to long-term database
Keeping queries in memory for later new attempt
```

And when the queries are finally written, you see this:
```
Notice: Queries stored in long-term database: 44 (took 142.7 ms, last SQLite ID 53921217)
        Queries from earlier attempt(s) stored successfully
```

But your log doesn't show evidence of this kind of issue (at least not in the available lines - the debug log shows only a few lines of `FTL.log`).

**Comment by nazar-pc** [comment]
Is there a more detailed log level that I can enable or something else of this kind?

**Comment by DL6ER** [comment]
This is an interesting case. Naively thinking, I guess every added debug logging would worsen your situation somewhat as all of Pi-hole's debug options are rather verbose.

`FTL.log` is barely ever written to and logging to `pihole.log` happens completely asynchronous. Following *The Linux Programming Interface: A Linux and UNIX System Programming Handbook*:

> the kernel buffer cache ensures that I/O on regular files does not block, as described in Section 13.1

If this would not be working, something in the kernel would be broken. We are not `sync`ing the log at any point and let the kernel itself handle things alone.

Storing to the database is furthermore done in a different *thread* altogether. It should not cause slowdowns to DNS resolution.

---snip---

Well, having had to make a break from typing here and returning later to the computer actually made me realize there is *one* possible issue I could imagine here: When writing to the database, we are temporarily locking the datastructure holding DNS queries to ensure consistent writes. However, we are needlessly locking it for longer than we need to - so far this has never been an issue but maybe it is in your case.

Fortunately, it's sufficiently easy to find out if my theory is right or not. Please run
```
pihole checkout ftl tweak/unlock_commit
```
and either find that the situation (greatly) improves or that nothing changes.

**Comment by nazar-pc** [comment]
Running it already, will report back after some time with it

**Comment by nazar-pc** [comment]
Observations so far: I do see a few seconds page load delays occasionally (which may or may not be related to DNS, didn't confirm yet), however there wasn't a single time that page didn't load at all.

Before switching to that branch DSN resolution was quickly failing outright, so there is definitely a visible positive change.

**Comment by nazar-pc** [comment]
The thing that broke though is deleting diagnostic messages. I'm getting:
```
Error, something went wrong!
While executing message statement: database is locked
```

**Comment by DL6ER** [comment]
> Observations so far: I do see a few seconds page load delays occasionally (which may or may not be related to DNS, didn't confirm yet), however there wasn't a single time that page didn't load at all.
> 
> Before switching to that branch DSN resolution was quickly failing outright, so there is definitely a visible positive change. 

This is great, I will open a PR to merge this into the [currently running Pi-hole v6.0 beta](https://pi-hole.net/blog/2023/10/09/pi-hole-v6-beta-testing/#page-content). There won't be another v5 release (this may or not change, but it is rather unlikely).

---

> I'm getting:
> ```
> Error, something went wrong!
> While executing message statement: database is locked
> ```

This is expected and cannot be avoided unless messages would be stored in a different database. This would increase complexity quite a bit and the solution should much more likely be running Pi-hole on a system with less load. It doesn't appear to be a good idea overall to run your Pi-hole at all on such a special system. Why are you running `sync` at all? Power-outage prevention? To my understanding, the `noop` sheduler will eventually write all files to disk on its own (over time), there will just no re-ordering happen (which is useful on SSDs). `sync` isn't strictly needed here.

---

> Considering how small gravity database is (8.7M in my case), it would make a lot of sense to keep its copy in memory, something under 50M will not cause issues for most users, while performance will improve massively and will avoid issue described here completely.

Sorry, I have missed this before. We are not talking about `gravity.db` but `pihole-FTL.db` which will typically be in the low GB range for most users. We cannot afford putting this into memory. The gravity database is indeed held in the kernel's page cache, it's always quickly available and shouldn't affect performance in your case. The optimization I have done in the proposed branch you tried improves committing to `pihole-FTL.db` alone.

**Comment by nazar-pc** [comment]
> This is expected and cannot be avoided unless messages would be stored in a different database.

This is expected and cannot be avoided unless messages would be stored in a different database. This would increase complexity quite a bit and the solution should much more likely be running Pi-hole on a system with less load.

> It doesn't appear to be a good idea overall to run your Pi-hole at all on such a special system. Why are you running sync at all? Power-outage prevention? To my understanding, the noop sheduler will eventually write all files to disk on its own (over time), there will just no re-ordering happen (which is useful on SSDs). sync isn't strictly needed here.

That was just one of the ways to reproduce the issue. I was having other workloads running there that caused high disk I/O as well. Already planning to upgrade the drive today or tomorrow, but since I did see this odd behavior I thought I better report it upstream and maybe something can be improved as the result.

> Sorry, I have missed this before. We are not talking about gravity.db but pihole-FTL.db which will typically be in the low GB range for most users. We cannot afford putting this into memory. The gravity database is indeed held in the kernel's page cache, it's always quickly available and shouldn't affect performance in your case. The optimization I have done in the proposed branch you tried improves committing to pihole-FTL.db alone.

I see. Is some in-memory buffer a viable option though? I don't think it needs to be synchronously written to disk when DNS request comes in. Writing records in bulk might also speed thing up a bit.

**Comment by DL6ER** [comment]
> Already planning to upgrade the drive today or tomorrow, but since I did see this odd behavior I thought I better report it upstream and maybe something can be improved as the result.

:+1: definitely and thank you for this! I checked the code of the currently running Pi-hole v6.0 beta just now and see that we did already make a similar optimization having the same effect that what you tried now.

> Is some in-memory buffer a viable option though? I don't think it needs to be synchronously written to disk when DNS request comes in. Writing records in bulk might also speed thing up a bit.

Yes, yes and yes.

- in-memory buffer: already used and ....
- it is already fully asynchronous - however, as we use low-level primitives to get best performance, when you manually call `sync` you **force** this buffer to the disk database **now** effectively removing the asynchronous aspect until your `sync` call returned. It is again asynchronous thereafter
- in bulk: by default we collect DNS queries for one minute and then write them in one batch to the database. The interval can be adjusted using the config option [`DBINTERVAL=1`](https://docs.pi-hole.net/ftldns/configfile/#dbinterval) in v5 and [`database.DBinterval=60`](https://github.com/pi-hole/FTL/blob/db32e4da4cd4eafafa1fb486027a8c745eeb6e59/src/config/config.c#L848-L852) in Pi-hole v6.0+

**Comment by nazar-pc** [comment]
> it is already fully asynchronous - however, as we use low-level primitives to get best performance, when you manually call sync you force this buffer to the disk database now effectively removing the asynchronous aspect until your sync call returned. It is again asynchronous thereafter

Makes sense, I just though about some in-app buffer rather than OS caching/buffering to have better control over it and make sure blocking is not happening in Pi-hole no matter what. Looking at it from the perspective that it is better to lose logs than to impact DNS resolution performance.

Either way glad to see improvements and looking forward to 6.0!

**Comment by nazar-pc** [comment]
I think I'm also sometimes hitting a similar issue under different circumstances.

If Internet goes down briefly Pi-hole stops responding for maybe a minute (gives out SERVFAIL in nslookup). Like it goes into some kind of defensive mode or something. Then a minute later it works fine again :man_shrugging:

**Comment by nazar-pc** [comment]
The issue is still very present on `2024.06.0` Docker release. If anything it just stops responding for a while. The expectation remains that it should be able to work in-memory in the meantime and not block DNS resolution.

**Comment by PromoFaux** [comment]
As you're running the Docker container, I might suggest trying the `development-v6` tag. Although currently in beta, it's in a very mature state at the moment (I've been running it since it existed). Would be interesting to see if you still experience the same/similar issues.

Forewarning - have a read over the README on that branch, and make a backup of any volumes first - there are a few changes from the v5 images

**Comment by nazar-pc** [comment]
Will give it a try over weekend, thank you!

**Comment by rdwebdesign** [comment]
[Docker Pi-hole **development-v6 README**](https://github.com/pi-hole/docker-pi-hole/tree/development-v6)

**Comment by nazar-pc** [comment]
Is there a migrarion from current stable version implemented at all?

I'm getting following errors:
```
pihole-1  | 2024-06-29 17:43:23.064 EEST [151M] ERROR: SQLite3: no such table: main.antigravity in "SELECT adlist_id from vw_antigravity WHERE domain = ? AND group_id IN (0);" (1)
pihole-1  | 2024-06-29 17:43:23.064 EEST [151M] ERROR: gravityDB_open("SELECT(... vw_antigravity ...)") - SQL error prepare: SQL logic error
pihole-1  | 2024-06-29 17:43:23.065 EEST [151M] ERROR: Gravity database not available (allowlist)
pihole-1  | 2024-06-29 17:43:24.273 EEST [174/F151] ERROR: SQLite3: no such table: main.antigravity in "SELECT adlist_id from vw_antigravity WHERE domain = ? AND group_id IN (0);" (1)
pihole-1  | 2024-06-29 17:43:24.273 EEST [174/F151] ERROR: gravityDB_open("SELECT(... vw_antigravity ...)") - SQL error prepare: SQL logic error
pihole-1  | 2024-06-29 17:43:24.273 EEST [174/F151] ERROR: Gravity database not available (allowlist)
```

Also I can't log in anymore and `docker compose run pihole pihole setpassword X` simply ran pihole again.

It complains about `FTLCONF_LOCAL_IPV4` even though it is still how readme suggests to set configuration option and it is still mentioned in https://github.com/pi-hole/docs/blob/release/v6.0/docs/ftldns/configfile.md#local_ipv4-unset-by-default-pr-1293-local_ipv4-data-toc-labelforce-local-a-reply

Will roll back to stable for now. Upgrade guide would have been nice to have.

**Comment by PromoFaux** [comment]
> mentioned in [pi-hole/docs@`release`/v6.0/docs/ftldns/configfile.md#local_ipv4-unset-by-default-pr-1293-local_ipv4-data-toc-labelforce-local-a-reply](https://github.com/pi-hole/docs/blob/release/v6.0/docs/ftldns/configfile.md?rgh-link-date=2024-06-29T14%3A47%3A11Z#local_ipv4-unset-by-default-pr-1293-local_ipv4-data-toc-labelforce-local-a-reply)

Oh my - I guess this page of the docs is not updated for v6, as that link is for the v5 `.conf` file, not the v6 `.toml` file. (@DL6ER is this still a WIP?) 

> `docker compose run pihole pihole setpassword X`

close, the command you're looking for is as below (Or you can set it via env var with FTLCONF_webserver_api_password

```
adam@adam-pc:~/repos/play$ docker exec -it pihole pihole setpassword X
  [✓] New password set
```
> Upgrade guide would have been nice to have.

Hear that, will look into something here.

> I'm getting following errors:

What did your compose file look like to be throwing these errors? Did you make any changes at all from the v5 version?

**Comment by nazar-pc** [comment]
> What did your compose file look like to be throwing these errors? Did you make any changes at all from the v5 version?

Didn't find anything in the docs that was relevant, so just updated tag:
```yaml
services:
  pihole:
    image: pihole/pihole:2024.06.0
    environment:
      TZ: 'Timezone'
      FTLCONF_LOCAL_IPV4: 192.168.x.y
    volumes:
      - "./etc-pihole:/etc/pihole"
      - "./etc-dnsmasq.d:/etc/dnsmasq.d"
    ports:
      - "192.168.x.y:53:53/tcp"
      - "192.168.x.y:53:53/udp"
      - "192.168.x.y:80:80/tcp"
```

**Comment by nazar-pc** [comment]
Well, I decreased load on the disk to address this, but it would be nice to address it upstream instead. No idea if it is better in v6 because I have no idea how to migrate to it.

**Comment by DL6ER** [comment]
See https://pi-hole.net/blog/2023/10/09/pi-hole-v6-beta-testing/#page-content

**Comment by nazar-pc** [comment]
That link was already shared above, @DL6ER, but it doesn't have migration guide for users with existing installations and as described above even after trying I didn't quite figure out what do I need to do to make it work. My setup is pretty unsophisticated.

**Comment by DL6ER** [comment]
The instructions describe how to do a fresh install + then followed by instructions how to upgrade a then existing v5 installation to v6.

We're close to merging the v6 beta code into regular development, though. Once done, a
```
pihole checkout dev
```
should be sufficient.

Currently, it is the
```
echo "development-v6" | sudo tee /etc/pihole/ftlbranch 
pihole checkout core development-v6
pihole checkout web development-v6
```
part of the link above.

**Comment by nazar-pc** [comment]
I'm running a Docker container. I expected to simply pull a new image, adjust the variables and it'll just work. I did try to update an image and described experience with broken installation above and rolled back to where it was. And I'm yet to find out what `FTLCONF_LOCAL_IPV4` corresponds to in v6.

**Comment by DL6ER** [comment]
The link also has

> The safest way to just try this out is using Docker and the `:development-v6` tag.

I think the corresponding options are `dns.reply.host.force4 = true` and `dns.reply.v4 = ...`

**Comment by nazar-pc** [comment]
I have latest `development-v6` a try again.

Here is the modified `docker-compose.yml`:
```yaml
services:
  pihole:
    image: pihole/pihole:development-v6
    environment:
      TZ: 'Timezone'
      FTLCONF_dns_reply_host_force4: 'true'
      FTLCONF_dns_reply_host_IPv4: 192.168.x.y
    volumes:
      - "./etc-pihole:/etc/pihole"
      - "./etc-dnsmasq.d:/etc/dnsmasq.d"
    ports:
      - "192.168.x.y:53:53/tcp"
      - "192.168.x.y:53:53/udp"
      - "192.168.x.y:80:80/tcp"
```

And here are the full logs after upgrade:
<details>

```
pihole-1  |   [i] Starting docker specific checks & setup for docker pihole/pihole
pihole-1  | 
pihole-1  |   [i] Ensuring basic configuration by re-running select functions from basic-install.sh
pihole-1  |   [i] No DNS upstream set in environment or config file, defaulting to Google DNS
pihole-1  |   [i] No password set in environment or config file, assigning random password: FPsRLn61
pihole-1  |    Upgrading gravity database from version 15 to 16
pihole-1  |    Upgrading gravity database from version 16 to 17
pihole-1  |    Upgrading gravity database from version 17 to 18
pihole-1  |    Upgrading gravity database from version 18 to 19
pihole-1  |   [i] Docker start setup complete
pihole-1  | 
pihole-1  |   [i] pihole-FTL (no-daemon) will be started as pihole
pihole-1  | 
pihole-1  |   [i] Setting capabilities on pihole-FTL where possible
pihole-1  |   [i] Applying the following caps to pihole-FTL:
pihole-1  |         * CAP_CHOWN
pihole-1  |         * CAP_NET_BIND_SERVICE
pihole-1  |         * CAP_NET_RAW
pihole-1  | 
pihole-1  | 2024-08-24 23:54:13.303 EEST [8M] INFO: Config file /etc/pihole/pihole.toml not available (r): No such file or directory
pihole-1  | 2024-08-24 23:54:13.326 EEST [14M] INFO: Config file /etc/pihole/pihole.toml not available (r): No such file or directory
pihole-1  | 2024-08-24 23:54:20.643 EEST [157M] INFO: ########## FTL started on 85585ef2993d! ##########
pihole-1  | 2024-08-24 23:54:20.643 EEST [157M] INFO: FTL branch: development-v6
pihole-1  | 2024-08-24 23:54:20.643 EEST [157M] INFO: FTL version: vDev-f40d895
pihole-1  | 2024-08-24 23:54:20.643 EEST [157M] INFO: FTL commit: f40d8958
pihole-1  | 2024-08-24 23:54:20.643 EEST [157M] INFO: FTL date: 2024-08-20 21:30:10 +0200
pihole-1  | 2024-08-24 23:54:20.643 EEST [157M] INFO: FTL user: pihole
pihole-1  | 2024-08-24 23:54:20.643 EEST [157M] INFO: Compiled for linux/amd64 (compiled on CI) using cc (Alpine 13.2.1_git20240309) 13.2.1 20240309
pihole-1  | 2024-08-24 23:54:20.645 EEST [157M] INFO: 2 FTLCONF environment variables found (2 used, 0 invalid, 0 ignored)
pihole-1  | 2024-08-24 23:54:20.645 EEST [157M] INFO:    [✓] FTLCONF_dns_reply_host_force4 is used
pihole-1  | 2024-08-24 23:54:20.645 EEST [157M] INFO:    [✓] FTLCONF_dns_reply_host_IPv4 is used
pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO: Wrote config file:
pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO:  - 148 total entries
pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO:  - 143 entries are default
pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO:  - 5 entries are modified
pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO:  - 2 entries are forced through environment
pihole-1  | 2024-08-24 23:54:20.647 EEST [157M] INFO: Parsed config file /etc/pihole/pihole.toml successfully
pihole-1  | 2024-08-24 23:54:20.647 EEST [157M] WARNING: Insufficient permissions to set process priority to -10 (CAP_SYS_NICE required), process priority remains at 0
pihole-1  | 2024-08-24 23:54:20.648 EEST [157M] INFO: PID of FTL process: 157
pihole-1  | 2024-08-24 23:54:20.649 EEST [157M] INFO: listening on 0.0.0.0 port 53
pihole-1  | 2024-08-24 23:54:20.649 EEST [157M] INFO: listening on :: port 53
pihole-1  | 2024-08-24 23:54:20.650 EEST [157M] INFO: PID of FTL process: 157
pihole-1  | 2024-08-24 23:54:20.652 EEST [157M] INFO: Database version is 12
pihole-1  | 2024-08-24 23:54:20.652 EEST [157M] INFO: Updating long-term database to version 13
pihole-1  | 2024-08-24 23:54:20.671 EEST [157M] INFO: Updating long-term database to version 14
pihole-1  | 2024-08-24 23:54:20.687 EEST [157M] INFO: Updating long-term database to version 15
pihole-1  | 2024-08-24 23:54:20.703 EEST [157M] INFO: Updating long-term database to version 16
pihole-1  | 2024-08-24 23:54:20.720 EEST [157M] INFO: Updating long-term database to version 17
pihole-1  | 2024-08-24 23:54:20.736 EEST [157M] INFO: Updating long-term database to version 18
pihole-1  | 2024-08-24 23:54:20.753 EEST [157M] INFO: Updating long-term database to version 19
pihole-1  | 2024-08-24 23:54:20.769 EEST [157M] INFO: Database successfully initialized
pihole-1  | 2024-08-24 23:54:22.075 EEST [157M] INFO: Imported 80847 queries from the on-disk database (it has 853180 rows)
pihole-1  | 2024-08-24 23:54:22.076 EEST [157M] INFO: Parsing queries in database
pihole-1  | 2024-08-24 23:54:22.162 EEST [157M] INFO:   10000 queries parsed...
pihole-1  | 2024-08-24 23:54:22.227 EEST [157M] INFO:   20000 queries parsed...
pihole-1  | 2024-08-24 23:54:22.287 EEST [157M] INFO:   30000 queries parsed...
pihole-1  | 2024-08-24 23:54:22.346 EEST [157M] INFO:   40000 queries parsed...
pihole-1  | 2024-08-24 23:54:22.407 EEST [157M] INFO:   50000 queries parsed...
pihole-1  | 2024-08-24 23:54:22.483 EEST [157M] INFO:   60000 queries parsed...
pihole-1  | 2024-08-24 23:54:22.551 EEST [157M] INFO:   70000 queries parsed...
pihole-1  | 2024-08-24 23:54:22.657 EEST [157M] INFO:   80000 queries parsed...
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO: Imported 80847 queries from the long-term database
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> Total DNS queries: 80847
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> Cached DNS queries: 6429
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> Forwarded DNS queries: 59502
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> Blocked DNS queries: 7534
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> Unknown DNS queries: 2
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> Unique domains: 1735
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> Unique clients: 3
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> DNS cache records: 3905
pihole-1  | 2024-08-24 23:54:22.670 EEST [157M] INFO:  -> Known forward destinations: 2
pihole-1  | 2024-08-24 23:54:22.687 EEST [157M] WARNING: Insufficient permissions to set system time (CAP_SYS_TIME required), NTP client not available
pihole-1  | 2024-08-24 23:54:22.688 EEST [157/T158] INFO: NTP server listening on 0.0.0.0:123 (IPv4)
pihole-1  | 2024-08-24 23:54:22.688 EEST [157/T159] INFO: NTP server listening on :::123 (IPv6)
pihole-1  | 2024-08-24 23:54:22.688 EEST [157M] INFO: FTL is running as user pihole (UID 100)
pihole-1  | 2024-08-24 23:54:22.691 EEST [157M] INFO: Created SSL/TLS certificate for pi.hole at /etc/pihole/tls.pem
pihole-1  | 2024-08-24 23:54:22.691 EEST [157M] INFO: Reading certificate from /etc/pihole/tls.pem ...
pihole-1  | 2024-08-24 23:54:22.691 EEST [157M] INFO: Using SSL/TLS certificate file /etc/pihole/tls.pem
pihole-1  | 2024-08-24 23:54:22.692 EEST [157M] INFO: Restored 0 API sessions from the database
pihole-1  | 2024-08-24 23:54:22.698 EEST [157M] INFO: Blocking status is enabled
pihole-1  | 2024-08-24 23:54:22.790 EEST [157/T160] INFO: Compiled 0 allow and 0 deny regex for 3 clients in 0.2 msec
pihole-1  | 2024-08-24 23:54:24.692 EEST [157M] ERROR: SQLite3: no such table: main.antigravity in "SELECT adlist_id from vw_antigravity WHERE domain = ? AND group_id IN (0);" (1)
pihole-1  | 2024-08-24 23:54:24.692 EEST [157M] ERROR: gravityDB_open("SELECT(... vw_antigravity ...)") - SQL error prepare: SQL logic error
pihole-1  | 2024-08-24 23:54:29.083 EEST [157M] WARNING: WARNING in dnsmasq core: ignoring query from non-local network 192.168.x.1 (logged only once)
```

</details>

The result:
* DNS doesn't work
* it doesn't let me in with previous credentials
* no indication of any upgrading attempt in logs
* SQL error towards the

So everything just like when I was getting when I tired v6 previously almost 2 months ago, except I was able to set two of the mentioned DNS options properly in environment variables now, so that is a progress I guess.

**Comment by DL6ER** [comment]
> DNS doesn't work

Not at all? Does it work for local names like `pi.hole` or `localhost`?

It may be caused by the warning
```
pihole-1  | 2024-08-24 23:54:29.083 EEST [157M] WARNING: WARNING in dnsmasq core: ignoring query from non-local network 192.168.x.1 (logged only once)
```
which is protecting your Pi-hole against being abused as an open-resolver. This is (and has been in v5) configured through the listening mode setting. On Pi-hole v6.0, the config is now called `dns.listeningMode` and you probably want to set its value to something like `ALL` but keep in mind that this should never be done on a server facing the public Internet without a proper firewall (even if this doesn't directly apply to you, I'm writing this for future readers).

---

> it doesn't let me in with previous credentials

@PromoFaux The log output above has

---

> `[i] No password set in environment or config file, assigning random password: FPsRLn61`

something that needs fixing?

---

> no indication of any upgrading attempt in logs

At the top of the logs you've quoted you can see all that.

---

> SQL error towards the [end]

Yes, this is due to the missing table `antigravity`. It is only created during a run of `pihole -g` - @PromoFaux something docker needs to add when it detects that this is a v5->v6 upgrade.

**Comment by rdwebdesign** [comment]
> [i] No password set in environment or config file, assigning random password: FPsRLn61

I think this is expected.
No password was set by environment variables, so a password was created for the web interface.

**Comment by nazar-pc** [comment]
> This is (and has been in v5) configured through the listening mode setting

And I had it configured in v5 that way indeed, I guess it is also one of the things that were not migrated from v5 automatically.

> At the top of the logs you've quoted you can see all that.

I guess it does some migration for some things, but this indicates it doesn't actually read the config:
> No DNS upstream set in environment or config file, defaulting to Google DNS

Because I did have DNS servers configured in v5 installation.

--

I guess initial setup of v6 is less bumpy, but upgrade process could use some work. I do not see anything except gravity database migrated according to logs.

**Comment by DL6ER** [comment]
> upgrade process could use some work

Migration did happen:

> ```
> pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO: Wrote config file:
> pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO:  - 148 total entries
> pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO:  - 143 entries are default
> pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO:  - 5 entries are modified
> pihole-1  | 2024-08-24 23:54:20.646 EEST [157M] INFO:  - 2 entries are forced through environment
> ```

see the last two lines here. Could you run `grep "-C10 CHANGED, " /etc/pihole/pihole.toml` so see which config values have been migrated? Is there anything missing you'd have expected being migrated? Maybe we add some more verbosity during migration to explicitly log which settings have been migrated from v5 -> v6.

**Comment by PromoFaux** [comment]
> something that needs fixing?

Yeah - the issue here was/is that the container is creating a `pihole.toml` file before FTL has had a chance to migrate the old configurations. My assumption here is that FTL says "I already have a config file, therefore the migration must already have happened"

https://github.com/pi-hole/docker-pi-hole/pull/1629 remedies this by allowing FTL to run and perform it's migrations _before_ we attempt to create/touch `pihole.toml`

**Comment by nazar-pc** [comment]
Just upgraded to `development` tag (`development-v6` seems to be gone). Password and other settings were transferred correctly, will see how it performs even though after optimizations on my end I have not had huge stalls recently.

UI seems way snappier now, very nice!

**Comment by nazar-pc** [comment]
Closing for now, didn't notice this any time recently after switching to dev versions of v6

**Comment by barrelltitor** [comment]
Happening here as well. I end up restarting the container and it works fine after. 

Lots of dns requests will end up with no result causing functionality all over the place to break

**Comment by nazar-pc** [comment]
It seems to get unstuck on its own for me. Well, it was, until I switched SSD and no longer had this problem.

It was happening with some problematic SSDs like Samsung 990 Pro, which has well known bug in the firmware, which I believe in addition to sometimes causing SSD to disappear from the system completely, causes regular stutters, which Pi-hole appears to be especially sensitive to for some reason.

Whatever the reason for it is, there should be a way to provide a path for DNS resolution that is completely independent from the disk activity with both an app and data necessary being already in memory.

**Comment by sgreadly** [comment]
For what it's worth, I found this while searching for the same symptoms. I did suffer from this for at least ~1.5 years or so, and it was still reproducible on v6.

Host: Raspberry Pi 5, 4GB. Pihole installed on bare metal, on an SSD via a [USB3 SATA adapter w/UASP](https://www.amazon.com.au/dp/B00HJZJI84).

```
$ pihole -v
Core version is v6.0.6 (Latest: v6.0.6)
Web version is v6.1 (Latest: v6.1)
FTL version is v6.1 (Latest: v6.1)


$ uname -a
Linux 6.12.20+rpt-rpi-2712 #1 SMP PREEMPT Debian 1:6.12.20-1+rpt1~bpo12+1 (2025-03-19) aarch64 GNU/Linux
```

My RPi is used for my pihole, as well as to download some files then move them to my NAS share via NFS.

I noticed the Pihole does not respond to DNS queries, either due to high disk I/O, or high network I/O. The issue is observed while I'm copying files from my RPi to my NFS mount, until the file transfer is complete, then I get the responses again. 

However, I switched to an M.2 (Argon One case) over the last 3-4 months and have not observed this issue any more.

**Comment by PayBas** [comment]
Decided to run Pi-hole on a Synology NAS with only HDDs (no SSDs), which means the OS and Pi-hole inside a Docker container also run on the HDDs.
When performing any serious disk operations (copying files, etc.) which generate moderately high disk utilization, Pi-hole becomes completely unresponsive and fails to reply to DNS requests.

```yaml
services:
  pihole:
    image: pihole/pihole:latest
    container_name: pihole-host
    cap_add:
      - CAP_NET_RAW
      - CAP_NET_BIND_SERVICE
      - CAP_CHOWN
      - SYS_NICE
    environment:
      - PIHOLE_UID=1031
      - PIHOLE_GID=65537
      - FTLCONF_webserver_port=8000
      - FTLCONF_webserver_api_password=xxx
      - FTLCONF_dns_listeningMode=local
      - DNSMASQ_USER=pihole
    volumes:
      - /volume1/docker/pihole/dnsmasq.d:/etc/dnsmasq.d
      - /volume1/docker/pihole/pihole:/etc/pihole
    network_mode: host
    ports:
      - 53:53/udp
      - 8000:8000/tcp
    restart: unless-stopped
```
Docker Tag [2025.08.0](https://github.com/pi-hole/docker-pi-hole/releases/2025.08.0) Core [v6.1.4](https://github.com/pi-hole/pi-hole/releases/v6.1.4) FTL [v6.2.3](https://github.com/pi-hole/FTL/releases/v6.2.3) Web interface [v6.2.1](https://github.com/pi-hole/web/releases/v6.2.1)

It's doesn't appear to be the CPU/RAM, since those barely break 20% load. But disk utilization any higher than 75% and DNS requests fail.

**Comment by runob** [comment]
Same problem for me PayBas, Synology DS418play running DSM 6.2.4. Issue started happening once migrated to v6. On v5 it always ran fine.

**Comment by PayBas** [comment]
@runob I'm on DSM 7.2.2-72806 on the exact same NAS model. I was hoping to run Pi-hole on the NAS so I wouldn't have to manage an entire dedicated device such as a RPi just for Pi-hole, but the experience has been pretty bad so far due to this issue.

**Comment by Peppe21** [comment]
This problem became prevalent since v.6 for me (ver. 2025.11.0), on Synology 7.1.1 via docker, under elevated I/O (backups via SMB). 
There was no problem in V.5, Now dns resolver time out systematically.
Had to apply a workaround to limit my backup bandwith to keep pi-hole responding...

**Comment by jimb0b** [comment]
Same problem here on a pi zero 2 w. I was getting frequent DNS timeouts from pihole and seems to be disk i/o related. 

I changed `database.DBinterval` to `3600` and now I only get DNS timeouts once an hour, which is slightly more tolerable(!)

**Comment by sdcarpenter** [comment]
Can confirm that I also am experiencing this on a Synology NAS. All I need to do to repro is to initiate a SMB copy and DNS resolution begins to fail.

**Comment by sgreadly** [comment]
So there have been numerous updates above with many people experiencing the same issue since this issue was opened ~1.5 years ago. Clearly there's an issue here.

Is any dev going to look into this now? Or do we need to keep adding more "yep, I see this too!" messages for the next 1.5 years?

**Comment by dschaper** [comment]
Seems to me that you fixed the problem, at least that's what you claimed previously:

> However, I switched to an M.2 (Argon One case) over the last 3-4 months and have not observed this issue any more.

Doing any highly resource intensive operation like writing to a networked filesystem is going to cause delays. Pi-hole runs with a lower `nice` level so that it is not the one starving the system of resources. My suggestion is to not run a vital network application like DNS on a shared system with limited resource that you regularly spike the IOPS and put everything in to contention. CPU not spiking tells me that the starvation is in the IOPS, which a spinning disk based NAS is going to be limited by. The Argon One has NVMe/SSD which drastically changes the situation with regards to IOPS.

**Comment by nazar-pc** [comment]
@dschaper that all makes sense, but as discussed previously, there should be no reason for DNS resolution to be blocked by the disk I/O at all. Disk I/O should happen asynchronously in the background, leaving the most important path of domain resolution completely free from disk I/O. If it is caused by something like logging and buffer size is exceeded, I'd rather for log lines to be dropped completely than for DNS resolution to suffer.

I had these issues even with PCIe Gen 3 NVMe SSDs, depending on SSD quality and what is happening on the system. So this is definitely not specific to HDDs or NASes in general.

**Comment by dschaper** [comment]
Everything is going to be blocked by disk I/O. When you have a system that is maxed out and swapping then the kernel is going to be sitting in IOWait state and locked by the processes it is waiting for IO for. It doesn't matter if everything related to Pi-hole is in RAM (SHM) when the processor isn't processing. That's the metric you need to be looking at, `iotop` and the amount of time the CPU is doing nothing, waiting for IO. If you want to make FTL run at `nice`level that puts FTL at the top of the priority list then you can, there's a setting for that. But to assume that FTL is going to run at full tilt when the CPU is in IOWAIT is not correct.

The long and short of it is to not run your sole network DNS server on a system that regularly runs out of resources. 

<img width="645" height="337" alt="Image" src="https://github.com/user-attachments/assets/b8f3d6c7-c101-49c7-85ee-6c4c691dcd06" />

**Comment by sgreadly** [comment]
"When you have a system that is maxed out and swapping then ....."

System was (and is) not maxed out, or swapping. It's hardly being used for anything else, and this only happens (at least for me) when I copy files via SMB. And it only started happening not too long ago - I've had this on older Pihole versions and copying files via SMB with no issues. I also had it on older RPis (RPi 4, RPi 3b) with no issues whilst copying files too. Something around pihole v5 changed or happened that causes resolution to fail now with the above symptoms. 

And yes, I fixed the issue with the M.2, but that was a temporary workaround for testing purposes, I had to return that to the owner and continue with my device. That said, I will be swapping this again next year so hopefully no more impact, but it's still strange that if this were truly related to IOWAIT, why did it start happening recently (i.e. within the last 1.5 year or so?)

Ta.

**Comment by nazar-pc** [comment]
My system has 128G of RAM, only 10% of which is actively used, ~50G of it is not used for anything at all, even caches. The system is not overloaded in any other way either. It just is waiting for disk I/O on a particular disk where Pi-hole happens to be installed during active use. Most of the CPU cores are doing nothing, RAM and other resources are not a limiting factor, not even close.

I might be missing something, but I see no reason why pi-hole can't continue doing network I/O even if disk I/O blocked completely. It is a DNS server first and foremost, so the DNS functionality should follow the least obstructed path with the lowest latency possible. If it doesn't work as DNS server, it doesn't matter whether it does blocking or not anymore.

**Comment by dschaper** [comment]
Run `iotop` and see how much is being blocked by being in iowait state. Everything is a file in linux, network I/O is exactly the same as disk I/O. Network endpoints are file handles. This is just the nature of linux and a shared scheduling system. You are not running a real-time OS.

Run a system with 128G of ram and zero swap. See what happens when the kernel needs to swap out memory for any reason and there's no disk to swap to. Run a system with a terrabyte of ram and exhausted IOPS and see what happens. Just because there is ram doesn't mean that it will be used. 

Why do you have 50G or RAM doing nothing? Shouldn't everything be loaded in to that RAM until it's maxed out and then swapped? 

This isn't a problem with Pi-hole not doing it's job, it's a problem of the kernel not being able to use the resources needed for Pi-hole to do it's job.

As for Raspberry Pi devices, the network interface is on the USB bus, the same bus that external hard drives are attached to. Doing heavy network i/o along with heavy disk i/o like NFS or SMB means the USB bus is maxed trying to move things from the network to the drive, all on the same limited bus.

**Comment by nazar-pc** [comment]
I do run 128G of RAM with no SWAP. `iotop` IMO is completely irrelevant. Even if the particular disk stops completely under load, all other disks and other operations should be still functioning just fine.

> Why do you have 50G or RAM doing nothing? Shouldn't everything be loaded in to that RAM until it's maxed out and then swapped?

I have 61G of caches right now, but also 52G completely free. "Why" is irrelevant, this is just how my system behaves right now. The important thing is that I'm not running out of memory, not using SWAP, etc.:
```
               total        used        free      shared  buff/cache   available
Mem:           125Gi        12Gi        52Gi        17Mi        61Gi       113Gi
Swap:             0B          0B          0B
```

> This isn't a problem with Pi-hole not doing it's job, it's a problem of the kernel not being able to use the resources needed for Pi-hole to do it's job.

I'm inclined to believe it is a problem with Pi-hole though. When disks other than the one Pi-hole is using are at 100% utilization (mostly HDDs), Pi-hole has no issues with it. The problem is only present when the disk Pi-hole is installed on and stores logs to is under stress. To me this seems to indicate that Pi-hole is trying to use that disk during critical operations of domain name resolution, preventing them from making progress.

And my system is far from Raspberry PI, I'm running AMD EPYC 7302P on a server motherboard. If even such system with tens of gigabytes of free RAM and tens of idle logical CPU cores has issues, it is clearly the software that needs to improve here.

**Comment by JamesTX10** [comment]
I had the issue with a pi zero 2 as well. Moved my primary pihole to a pi 5
with SSD and no longer had issues. If it is fixed, I need to use the  pi
zero 2 for a secondary.

**Comment by sdcarpenter** [comment]
The most recent build seems to have fixed this!!! I ran a large SMB copy on my underpowered NAS and then ran a number of `nslookups` and they all returned successfully. This used to always fail.

**Comment by PayBas** [comment]
Can confirm. Running 2025.11.1 seems to have resolved this issue.

**Comment by Peppe21** [comment]
Same here issue resolved.

**Comment by dschaper** [comment]
I don't see anything that really changed in the new releases. Maybe something on the `sqlite3` end? 

I'm very glad things seem to be resolved, I would like to see some results of `iotop` of the versions affected and the versions resolved.

**Comment by jimb0b** [comment]
> I don't see anything that really changed in the new releases. Maybe something on the `sqlite3` end?
> 
> I'm very glad things seem to be resolved, I would like to see some results of `iotop` of the versions affected and the versions resolved.

Looks like these two to me (at least, I can also confirm my symptoms have resolved and were definitely worse during each database.DBinterval so suspect these PRs fixed my particular issue):

https://github.com/pi-hole/FTL/pull/2700
https://github.com/pi-hole/FTL/issues/2716
